compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -540,6 +540,39 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
             offset = offset.align_to(field_align.abi);
             align = align.max(field_align);
 
+            // If `-Z randomize-layout` is enabled, we pad each field by a multiple of its alignment
+            // If layout randomization is disabled, we don't pad it by anything and if it is
+            // we multiply the field's alignment by anything from zero to the user provided
+            // maximum multiple (defaults to three)
+            //
+            // When `-Z randomize-layout` is enabled that doesn't necessarily mean we can
+            // go ham on every type that at first glance looks valid for layout optimization.
+            // `Option` specifically has layout guarantees when it has specific `T` substitutions,
+            // such as `Option<NonNull<_>>` or `Option<NonZeroUsize>` both being exactly one `usize`
+            // large. As such, we have to ensure that the type doesn't guarantee niche optimization
+            // with the current payload
+            let extra_padding = Size::from_bytes(
+                if repr.can_randomize_type_layout()
+                    && !util::nonnull_opt_applies(self.tcx, ty, true)
+                {
+                    field_align
+                        .abi
+                        .bytes()
+                        .checked_mul(rng.gen_range(0..=repr.random_padding_max_factor))
+                        .unwrap_or(0)
+                } else {
+                    0
+                },
+            );
+
+            // Create our "normal" offset, without applying our extra padding
+            let unpadded_offset =
+                offset.checked_add(field.size, dl).ok_or(LayoutError::SizeOverflow(ty))?;
+
+            // Attempt to add our extra padding, defaulting to our unpadded offset
+            // in the case of an overflow
+            offset = unpadded_offset.checked_add(extra_padding, dl).unwrap_or(unpadded_offset);
```

Comment:
Hmm, hold on, this doesn't look right. Adding `field.size` at the very end is done to *move past* the field, but the value written by `offsets[i as usize] = offset;` has to be *the field's offset* (i.e. without `field.size` being added in).

(The change to the niche computation tipped me off, and before that I was curious why tuples started getting randomized despite there not being even something like a libcore lang item which could be used to determine how to randomize tuples based on flags passed when building libcore)

Reply:
Ahh ok I see, I think I fixed that now